include/internal/safe_math.h

```diff
@@ -292,6 +292,112 @@
         return a;                                                            \
     }
 
+/*
+ * Safe shift operations.
+ *
+ * The C standard defines shifts by negative values and shifts larger than
+ * the width of the type to be undefined behaviours.  We treat negative shifts
+ * as shifts in the other direction and too large shifts push everything out.
+ *
+ * Signed left shifts that move a 1 bit into or out of the sign bit are
+ * undefined.  Right shifts of negative values are implemenetation defined.
+ */
+
+/*
+ * Signed shift left fails if set bits are shifted into or out of the sign bit.
+ * Shifts zero bits into the lower order end.
+ */
+# define OSSL_SAFE_MATH_SHLS(type_name, type) \
+    static ossl_inline ossl_unused type safe_shr_ ## type_name(type a,      \
+                                                               int n,       \
+                                                               int *err);   \
+    static ossl_inline ossl_unused type safe_shl_ ## type_name(type a,      \
+                                                               int n,       \
+                                                               int *err)    \
+    {                                                                       \
+        if (n == 0)                                                         \
+            return a;                                                       \
+        if (n < 0)                                                          \
+            return safe_shr_ ## type_name(a, -n, err);                      \
+        if (n >= 8 * (int)sizeof(type)) {                                   \
+            *err |= a != 0;                                                 \
+            return 0;                                                       \
+        }                                                                   \
+        if ((a & ~((1 << ((int)sizeof(type) * 8 - n - 1)) - 1)) != 0) {     \
+            *err = 1;                                                       \
+            return 0;                                                       \
+        }                                                                   \
+        return a << n;                                                      \
+    }
+
+/*
+ * Signed shift right inserts zero bits into the top.
+ */
+# define OSSL_SAFE_MATH_SHRS(type_name, type) \
```

Comment:
I think it is confusing to name these shr and ashr for signed types. For signed types the default assumption is that a right-shift does sign extension, so someone who doesn't notice the existence of ashr is liable to be quite surprised by the behaviour of shr. I'd like to see shr named ushr.

I don't mind shr also being called ushr on unsigned types, for consistency. It doesn't hurt to be unambiguous when things like this are concerned.

I'm skeptical about the utility of allowing negative shift values and would prefer to see them disallowed and result in an error, at least for now.

Reply:
Arithmetic shift right versus plain shift right.  Why not use ASR for the name?  That's what all assembly languages I've used have called it (with some variations e.g. SAR on some machines).

I'm not keen on _ushr_, keep the names simple: the type is implicit in the function name so I don't see a need to further distinguish signed and unsigned.

I'm in partial agreement about negative shifts.  I'm okay either way, I'd prefer what's more useful and this is a very rare case IMO.  It's entirely possible to switch to an unsigned _n_ argument which avoids this entirely.

Reply:
My testing here found that negative right shifts introduced zero bits.

The C standard is quite clear saying that it is implementation defined.  It doesn't even provide alternatives that are legal so a right shift of a negative value could alternately insert 0 and 1 bits and still be conformant.

Reply:
To my mind if someone is passing a negative shift value this is likely to indicate a bug in their code. I can see very little compelling use for this and if someone genuinely does need to dynamically switch between shifting left and shifting right they can trivially do so manually. It's not like we're even talking about a roll operation here so a dynamic shift direction isn't very useful for crypto code.

Reply:
The type name remains in the function name.  I'm seeing zero benefit from `safe_ushr_uint32` versus `safe_shr_uint32`.  The unsigned is implicit in the type.

Yeah, negative shifts are dubious.  Easy to remove and error out.

Should zero shifts be treated likewise?  Again, more likely to be a programming error than wanted.

Reply:
I see no issue with zero shifts. IMO they should be allowed.

I'm talking about `safe_shr_int32` not `safe_shr_uint32`. If you have instantiated these functions for a signed type, making `safe_shr` be an unsigned shift is confusing.